lib/topological_inventory/ansible_tower/targeted_refresh/service_instance.rb

```diff
@@ -0,0 +1,202 @@
+        end
+
+        private
+
+        def refresh_part_cloud(tasks_id, refresh_state_uuid, refresh_state_part_uuid, query_params)
+          parser = TopologicalInventory::AnsibleTower::Parser.new(:tower_url => tower_hostname)
+
+          get_service_instances(connection, query_params).each do |service_instance|
+            parser.parse_service_instance(service_instance)
+          end
+
+          # Sending to Ingress API
+          logger.info("ServiceInstance#refresh - Task[ id: #{tasks_id} ] Sending to Ingress API...")
+          save_inventory(parser.collections.values, inventory_name, schema_name, refresh_state_uuid, refresh_state_part_uuid)
+          logger.info("ServiceInstance#refresh - Task[ id: #{tasks_id} ] Sending to Ingress API...Complete")
+        end
+
+        def connection
+          @connection ||= begin
+                            tower_user = authentication.username unless on_premise?
+                            tower_passwd = authentication.password unless on_premise?
+
+                            connection_manager.connect(
+                              :base_url       => tower_hostname,
+                              :username       => tower_user,
+                              :password       => tower_passwd,
+                              :receptor_node  => endpoint.receptor_node.to_s.strip,
+                              :account_number => account_number
+                            )
+                          end
+        end
+
+        def on_premise?
+          @on_premise ||= endpoint.receptor_node.to_s.strip.present?
+        end
+
+        def set_connection_data!
+          raise TopologicalInventory::Providers::Common::Operations::Source::ERROR_MESSAGES[:endpoint_not_found] unless endpoint
+          raise TopologicalInventory::Providers::Common::Operations::Source::ERROR_MESSAGES[:authentication_not_found] unless authentication
+
+          self.tower_hostname = full_hostname(endpoint)
+        end
+
+        # TODO: Join with other operations
+        def endpoint
+          @endpoint ||= api_client.fetch_default_endpoint(source_id)
+        rescue => e
+          logger.error("ServiceInstance#refresh - Failed to fetch Endpoint for Source #{source_id}: #{e.message}")
+          nil
+        end
+
+        def authentication
+          @authentication ||= api_client.fetch_authentication(source_id, endpoint)
+        rescue => e
+          logger.error("ServiceInstance#refresh - Failed to fetch Authentication for Source #{source_id}: #{e.message}")
+          nil
+        end
+
+        # Queries Sources API in the context of first task
+        def identity
+          @identity ||= params.to_a.first['request_context']
+        end
+
+        def api_client
+          @api_client ||= TopologicalInventory::Providers::Common::Operations::SourcesApiClient.new(identity)
+        end
+
+        # TODO: Join with PR #112 (ordering)
```

Comment:
Great idea!

Reply:
it'll wait to next PR because there are more duplicates accross operations